Minor refactoring of variable access

peter-b opened the pull request:
This set of patches follows on from #5260 with some further refactoring of LCS variable handling code.  The intention is to eventually refactor `MCVariable` in a similar way to the way `MCContainer` was changed in #5260 in order to be able to move `MCVariable` instances from the heap to the stack, or to coalesce `MCVariable` allocations into single bulk allocations on the heap.  Both these changes would improve cache locality when executing LCS code.

livecodeali commented:
@livecode-vulcan review ok f864e04

livecode-vulcan commented:
💙 review by @livecodeali ok f864e0413d22d47219df83f5e8211c791c142bf9

livecode-vulcan commented:
😞 test failure f864e0413d22d47219df83f5e8211c791c142bf9

* try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1133)
* try-community-mac: [failure](https://vulcan.livecode.com/builders/try-community-mac/builds/1135)
* try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1128)
* try-community-emscripten: [failure](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1134)
* try-community-ios: [failure](https://vulcan.livecode.com/builders/try-community-ios/builds/1138)
* try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1148)
* try-community-android-armv6: [failure](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1129)

livecodeali commented:
@livecode-vulcan review ok c00f985

livecode-vulcan commented:
💙 review by @livecodeali ok c00f98538e38cf463edbaa0a78408b91bfee635b

livecode-vulcan commented:
😞 test failure c00f98538e38cf463edbaa0a78408b91bfee635b

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/46)
* try-community-armv6-android-api9: [failure](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/45)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/32)
* try-community-universal-ios-iphoneos10.2: [failure](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.2/builds/50)
* try-community-universal-ios-iphonesimulator10.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.2/builds/50)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/44)
* try-community-universal-mac-macosx10.9: [failure](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/46)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/48)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/50)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/48)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/50)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/50)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/47)

peter-b committed:
Change MCContainer::m_path to managed pointer

Remove explicit deletion and span-construction code.

peter-b committed:
Remove std::move() call that prevents copy elision

peter-b committed:
Spanify MCVariable and MCContainer methods

Change all `MCVariable` and `MCContainer` methods that take a pointer
+ length pair but don't take ownership to use `MCSpan`.

peter-b committed:
libfoundation: Add span overloads for array accessors

Overload `MCArrayStoreValueOnPath()`, `MCArrayFetchValueOnPath()` and
`MCArrayRemoveValueOnPath()` with versions that take a name span
instead of a pointer + length.

peter-b committed:
Simplify construction of MCVariable, etc.

Provide static initialisers and use explicitly defaulted constructors,
where possible.

peter-b committed:
Remove unused MCHandler::getparamlist() method

runrevmark commented:
@livecode-vulcan review ok 89a24f9

livecode-vulcan commented:
💙 review by @runrevmark ok 89a24f9c22f4824aa08ae947cb16191263b066d7

livecode-vulcan commented:
😞 test failure 89a24f9c22f4824aa08ae947cb16191263b066d7

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/121)
* try-community-armv6-android-api9: [pending](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/121)
* try-community-js-emscripten-sdk1.35: [pending](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/107)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/53)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/53)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/119)
* try-community-universal-mac-macosx10.9: [pending](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/121)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/123)
* try-community-x86-linux-debian8: [pending](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/125)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/123)
* try-community-x86_64-linux-debian8: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/125)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/141)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/120)

peter-b commented:
@livecode-vulcan retry 89a24f9

livecode-vulcan commented:
😎 test success 89a24f9c22f4824aa08ae947cb16191263b066d7

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/123)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/123)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/109)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/55)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/55)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/121)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/123)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/125)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/127)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/125)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/127)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/143)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/122)
